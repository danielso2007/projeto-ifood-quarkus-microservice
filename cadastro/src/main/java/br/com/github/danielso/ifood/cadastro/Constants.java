package br.com.github.danielso.ifood.cadastro;

public final class Constants {
	
	private Constants() {}
	
	public static final String API_VERSION = "/v1";

}
