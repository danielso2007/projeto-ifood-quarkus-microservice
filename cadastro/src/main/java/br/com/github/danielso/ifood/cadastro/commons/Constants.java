package br.com.github.danielso.ifood.cadastro.commons;

public final class Constants {
	
	private Constants() {}
	
	public static final String API_VERSION = "/v1";
	public static final String REST_RESTAURANTE = "/restaurantes";
	public static final String REST_PRATO = "/pratos";

}
