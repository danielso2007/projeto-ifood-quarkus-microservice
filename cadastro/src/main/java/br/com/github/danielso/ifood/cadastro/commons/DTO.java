package br.com.github.danielso.ifood.cadastro.commons;

import javax.validation.ConstraintValidatorContext;

public interface DTO {

	default boolean isValid(ConstraintValidatorContext constraintValidatorContext) {
		return true;
	}

}
