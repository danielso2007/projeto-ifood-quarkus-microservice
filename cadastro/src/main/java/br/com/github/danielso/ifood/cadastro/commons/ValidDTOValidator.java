package br.com.github.danielso.ifood.cadastro.commons;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

public class ValidDTOValidator implements ConstraintValidator<ValidDTO, DTO> {

	@Override
	public void initialize(ValidDTO constraintAnnotation) {
		// Construtor padr√£o.
	}

	@Override
	public boolean isValid(DTO dto, ConstraintValidatorContext context) {
		return dto.isValid(context);
	}

}
