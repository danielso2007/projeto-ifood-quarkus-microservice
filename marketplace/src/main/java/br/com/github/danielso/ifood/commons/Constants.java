package br.com.github.danielso.ifood.commons;

public final class Constants {
	
	private Constants() {}
	
	public static final String API_VERSION = "/v1";
	public static final String REST_PRATOS = "/pratos";

}
