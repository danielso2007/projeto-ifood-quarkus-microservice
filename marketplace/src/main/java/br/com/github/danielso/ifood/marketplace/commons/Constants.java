package br.com.github.danielso.ifood.marketplace.commons;

public final class Constants {
	
	private Constants() {}
	
	public static final String API_VERSION = "/v1";
	public static final String REST_PRATOS = "/pratos";
	public static final String REST_RESTAURANTES = "/restaurantes";

}
