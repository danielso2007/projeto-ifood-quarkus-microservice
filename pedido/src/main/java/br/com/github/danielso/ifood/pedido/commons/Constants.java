package br.com.github.danielso.ifood.pedido.commons;

public final class Constants {
	
	private Constants() {}
	
	public static final String API_VERSION = "/v1";
	public static final String REST_PEDIDOS = "/pedidos";

}
