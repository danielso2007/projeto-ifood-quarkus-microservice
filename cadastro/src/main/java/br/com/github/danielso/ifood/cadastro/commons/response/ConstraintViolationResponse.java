package br.com.github.danielso.ifood.cadastro.commons.response;

import java.util.ArrayList;
import java.util.List;

import javax.validation.ConstraintViolationException;

import br.com.github.danielso.ifood.cadastro.commons.ConstraintViolationImpl;

public class ConstraintViolationResponse {

	private final List<ConstraintViolationImpl> errors = new ArrayList<>();

	private ConstraintViolationResponse(ConstraintViolationException exception) {
		exception.getConstraintViolations().forEach(error -> errors.add(ConstraintViolationImpl.of(error)));
	}

	public static ConstraintViolationResponse of(ConstraintViolationException exception) {
		return new ConstraintViolationResponse(exception);
	}

	public List<ConstraintViolationImpl> getErrors() {
		return errors;
	}

}
