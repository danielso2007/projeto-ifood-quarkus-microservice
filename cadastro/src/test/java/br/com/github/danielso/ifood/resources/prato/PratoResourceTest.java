package br.com.github.danielso.ifood.resources.prato;

import static io.restassured.RestAssured.given;

import java.util.Map;

import org.approvaltests.Approvals;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import br.com.github.danielso.ifood.cadastro.commons.Constants;
import br.com.github.danielso.ifood.testlifecyclemanager.CadastroTestLifecycleManager;
import io.quarkus.test.common.QuarkusTestResource;
import io.quarkus.test.junit.QuarkusTest;
import io.restassured.http.ContentType;
import io.restassured.response.Response;

@QuarkusTest
@QuarkusTestResource(CadastroTestLifecycleManager.class)
class PratoResourceTest {

	private static String requestBody = "{\n"
			+ "    \"descricao\": \"Feito com os melhores ingredientes\",\n"
			+ "    \"nome\": \"Macarrão ao molhor vermelho\",\n"
			+ "    \"preco\": 58.98\n"
			+ "  }";
	private static String requestBodyPrecoNulo = "{\n"
			+ "    \"descricao\": \"Feito com os melhores ingredientes\",\n"
			+ "    \"nome\": \"Macarrão ao molhor vermelho\" "
			+ "  }";
	private static String requestBodyNomeNulo = "{\n"
			+ "    \"descricao\": \"Feito com os melhores ingredientes\",\n"
			+ "    \"preco\": 58.98\n"
			+ "  }";
	
	@Test
	void testBuscarPratos_com_sucesso() throws JsonMappingException, JsonProcessingException {
		String resultado = given()
				.contentType(ContentType.JSON)
				.when()
				.get(Constants.API_VERSION + Constants.REST_RESTAURANTE + "/1/pratos")
				.then().statusCode(200)
				.extract()
				.asString();
		Approvals.verifyJson(ajustarDataParaZeroHorasEhMinutos(resultado));
	}
	
	@Test
	void testBuscarPrato_com_sort_e_order() throws JsonMappingException, JsonProcessingException {
		String resultado = given()
				.contentType(ContentType.JSON)
				.queryParam("sort", "id")
				.queryParam("order", "asc")
				.when()
				.get(Constants.API_VERSION + Constants.REST_RESTAURANTE + "/1/pratos")
				.then()
				.statusCode(200)
				.extract()
				.asString();
		Approvals.verifyJson(ajustarDataParaZeroHorasEhMinutos(resultado));
	}
	
	@Test
	void test_prato_find_by_id() throws JsonMappingException, JsonProcessingException {
		String resultado = given()
				.contentType(ContentType.JSON)
				.when()
				.get(Constants.API_VERSION + Constants.REST_RESTAURANTE + "/1/pratos/1")
				.then()
				.statusCode(200)
				.extract()
				.asString();
		Approvals.verifyJson(ajustarDataParaZeroHorasEhMinutos(resultado));
	}
	
	@Test
	void testBuscarPrato_com_sort_e_order_desc() throws JsonMappingException, JsonProcessingException {
		String resultado = given()
				.contentType(ContentType.JSON)
				.queryParam("sort", "id")
				.queryParam("order", "desc")
				.when()
				.get(Constants.API_VERSION + Constants.REST_RESTAURANTE + "/1/pratos")
				.then()
				.statusCode(200)
				.extract()
				.asString();
		Approvals.verifyJson(ajustarDataParaZeroHorasEhMinutos(resultado));
	}
	
	@Test
	void testBuscarPrato_com_sort_por_nome() throws JsonMappingException, JsonProcessingException {
		String resultado = given()
				.contentType(ContentType.JSON)
				.queryParam("sort", "nome")
				.queryParam("order", "desc")
				.when()
				.get(Constants.API_VERSION + Constants.REST_RESTAURANTE + "/1/pratos")
				.then()
				.statusCode(200)
				.extract()
				.asString();
		Approvals.verifyJson(ajustarDataParaZeroHorasEhMinutos(resultado));
	}

	@Test
	void testCadastroPrato() {
		Response response = given()
				.header("Content-type", "application/json")
				.and()
				.body(requestBody)
				.when()
				.post(Constants.API_VERSION + Constants.REST_RESTAURANTE + "/1/pratos")
				.then()
				.statusCode(201)
				.extract()
				.response();
		Assertions.assertEquals(201, response.statusCode());
	}
	
	@Test
	void testCadastroPrato_nome_nulo_erro_validacao() {
		Response response = given()
				.header("Content-type", "application/json")
				.and()
				.body(requestBodyNomeNulo)
				.when()
				.post(Constants.API_VERSION + Constants.REST_RESTAURANTE + "/1/pratos")
				.then()
				.statusCode(400)
				.extract()
				.response();
        
        Map<String, String> parameterViolations = response.jsonPath().getMap("errors[0]");
		
		Assertions.assertEquals(400, response.statusCode());
		Assertions.assertEquals("nome", parameterViolations.get("atributo"));
		Assertions.assertTrue(parameterViolations.get("mensagem").contains("O nome não pode ser "));
	}

	@Test
	void testCadastroPrato_preco_nulo_erro_validacao() {
		Response response = given()
				.header("Content-type", "application/json")
				.and()
				.body(requestBodyPrecoNulo)
				.when()
				.post(Constants.API_VERSION + Constants.REST_RESTAURANTE + "/1/pratos")
				.then()
				.statusCode(400)
				.extract()
				.response();
		
		Map<String, String> parameterViolations = response.jsonPath().getMap("errors[0]");
		
		Assertions.assertEquals(400, response.statusCode());
		Assertions.assertEquals("preco", parameterViolations.get("atributo"));
		Assertions.assertEquals("O preço não pode ser nulo", parameterViolations.get("mensagem"));
	}
	
	@Test
	void testAtualizarPrato() {
		Response response = given()
				.header("Content-type", "application/json")
				.and()
				.body(requestBody)
				.when()
				.put(Constants.API_VERSION + Constants.REST_RESTAURANTE + "/1/pratos" + "/1")
				.then()
				.extract()
				.response();
		Assertions.assertEquals(500, response.statusCode());
	}
	
	@Test
	void testAtualizarPrato_entidade_nao_existe_error() {
		Response response = given()
				.header("Content-type", "application/json")
				.and()
				.body(requestBody)
				.when()
				.put(Constants.API_VERSION + Constants.REST_RESTAURANTE + "/1/pratos" + "/999999")
				.then()
				.statusCode(404)
				.extract()
				.response();
		Assertions.assertEquals(404, response.statusCode());
	}
	
	@Test
	void testDeletandoPrato() {
		Response response = given()
				.header("Content-type", "application/json")
				.and()
				.body(requestBody)
				.when()
				.delete(Constants.API_VERSION + Constants.REST_RESTAURANTE + "/4/pratos" + "/3")
				.then()
				.extract()
				.response();
		Assertions.assertEquals(200, response.statusCode());
	}
	
	private String ajustarDataParaZeroHorasEhMinutos(String resultado) throws JsonMappingException, JsonProcessingException {
		ObjectMapper mapper = new ObjectMapper();
		JsonNode rootNode = mapper.readTree(resultado);
		if (rootNode.isArray()) {
			for (final JsonNode objNode : rootNode) {
		        ((ObjectNode)objNode).put("dataCriacao", "2021-01-01T03:00:00");
		        ((ObjectNode)objNode).put("dataAtualizacao", "2021-01-01T03:00:00");
		    }
		}
		return rootNode.toString();
	}
	
}
