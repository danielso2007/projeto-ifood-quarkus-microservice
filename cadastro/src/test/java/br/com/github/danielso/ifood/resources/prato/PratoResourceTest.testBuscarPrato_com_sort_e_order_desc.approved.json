[
  {
    "id": 30,
    "dataAtualizacao": "2021-01-01T00:00:00",
    "dataCriacao": "2021-01-01T00:00:00",
    "descricao": "Massa tincidunt sollicitudin semper, nibh eros imperdiet est, vitae pharetra mauris lacus quis est.",
    "nome": "Curabitur fermentum",
    "preco": 999.98
  },
  {
    "id": 20,
    "dataAtualizacao": "2021-01-01T00:00:00",
    "dataCriacao": "2021-01-01T00:00:00",
    "descricao": "Mauris euismod neque ut nibh scelerisque mollis. Pellentesque scelerisque, massa tincidunt sollicitudin semper, nibh eros imperdiet est, vitae pharetra mauris lacus quis est.",
    "nome": "Ut tempor ligula",
    "preco": 35.00
  },
  {
    "id": 10,
    "dataAtualizacao": "2021-01-01T00:00:00",
    "dataCriacao": "2021-01-01T00:00:00",
    "descricao": "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
    "nome": "Vestibulum at purus",
    "preco": 10.99
  }
]